farmerbot/internal/state.go

```diff
@@ -0,0 +1,272 @@
+	}
+
+	if inputs.Power.OverProvisionCPU < 1 || inputs.Power.OverProvisionCPU > 4 {
+		return nil, fmt.Errorf("cpu over provision should be a value between 1 and 4 not %v", inputs.Power.OverProvisionCPU)
+	}
+
+	// set farm
+	farm, err := sub.GetFarm(inputs.FarmID)
+	if err != nil {
+		return nil, err
+	}
+
+	// set nodes
+	nodes, err := convertInputsToNodes(ctx, sub, rmbNodeClient, inputs, farm.DedicatedFarm)
+	if err != nil {
+		return nil, err
+	}
+
+	config := state{
+		farm:   *farm,
+		nodes:  nodes,
+		config: inputs,
+	}
+
+	err = config.validate()
+	if err != nil {
+		return nil, err
+	}
+
+	return &config, nil
+}
+
+// ConvertInputsToNodes converts the config nodes from configuration inputs
+func convertInputsToNodes(ctx context.Context, sub Sub, rmbNodeClient RMB, config Config, dedicatedFarm bool) (map[uint32]node, error) {
+	nodes := make(map[uint32]node)
+
+	farmNodes, err := sub.GetNodes(config.FarmID)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, nodeID := range farmNodes {
+		if slices.Contains(config.ExcludedNodes, nodeID) {
+			continue
+		}
+
+		// if the user specified included nodes or
+		// no nodes are specified so all nodes will be added (except excluded)
+		if slices.Contains(config.IncludedNodes, nodeID) || len(config.IncludedNodes) == 0 {
+			neverShutDown := slices.Contains(config.NeverShutDownNodes, nodeID)
+
+			configNode, err := getNodeWithLatestChanges(ctx, sub, rmbNodeClient, nodeID, neverShutDown, false, dedicatedFarm)
+			if err != nil {
+				return nil, errors.Wrapf(err, "failed to include node with id %d", nodeID)
+			}
+			nodes[nodeID] = configNode
+		}
+	}
+
+	return nodes, nil
+}
+
+func getNodeWithLatestChanges(
+	ctx context.Context,
+	sub Sub,
+	rmbNodeClient RMB,
+	nodeID uint32,
+	neverShutDown,
+	hasClaimedResources,
+	dedicatedFarm bool,
+) (node, error) {
+
+	log.Debug().Uint32("ID", nodeID).Msg("Include node")
+	nodeObj, err := sub.GetNode(nodeID)
+	if err != nil {
+		return node{}, errors.Wrapf(err, "failed to get node %d from substrate", nodeID)
+	}
+
+	configNode := node{
+		Node: *nodeObj,
+	}
+
+	configNode.neverShutDown = neverShutDown
+
+	price, err := sub.GetDedicatedNodePrice(nodeID)
+	if err != nil {
+		return node{}, errors.Wrapf(err, "failed to get node %d dedicated price from substrate", nodeID)
+	}
+
+	if price != 0 || dedicatedFarm {
+		configNode.dedicated = true
+	}
+
+	rentContract, err := sub.GetNodeRentContract(nodeID)
+	if errors.Is(err, substrate.ErrNotFound) {
+		configNode.hasActiveRentContract = false
+	} else if err != nil {
+		return node{}, errors.Wrapf(err, "failed to get node %d rent contract from substrate", nodeID)
+	}
+
+	configNode.hasActiveRentContract = rentContract != 0
+
+	powerTarget, err := sub.GetPowerTarget(nodeID)
+	if err != nil {
+		return node{}, errors.Wrapf(err, "failed to get node %d power target from substrate", nodeID)
+	}
+
+	if powerTarget.State.IsUp && powerTarget.Target.IsUp && configNode.powerState != on {
```

Comment:
how the `configNode.powerState` will has any value if you just created configNode and never assigned anything to powerState

And since the zero value of powerState is ON it means this condition will never ever met